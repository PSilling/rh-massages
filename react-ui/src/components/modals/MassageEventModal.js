// react imports
import React, { Component } from "react";
import PropTypes from "prop-types";

// module imports
import { Row, Col, Button, Modal, ModalBody, Tooltip } from "reactstrap";
import moment from "moment";

// component imports
import ConfirmationIconButton from "../iconbuttons/ConfirmationIconButton";
import ModalActions from "../buttons/ModalActions";
import TooltipIconButton from "../iconbuttons/TooltipIconButton";

// util imports
import Auth from "../../util/Auth";
import _t from "../../util/Translations";
import Util from "../../util/Util";

/**
 * Modal displaying details about a given Massage event.
 * Also supports assignment management.
 */
class MassageEventModal extends Component {
  state = { tooltipActive: false };

  tooltipTarget = Util.getTooltipTargets(1)[0];

  definitionClass = "mb-3";

  onCalendarConfirm = () => {
    window.open(Util.getEventLink(this.props.event.massage), "_blank");
    this.props.onConfirm();
  };

  toggleTooltip = () => {
    this.setState(prevState => ({ tooltipActive: !prevState.tooltipActive }));
  };

  handleKeyPress = event => {
    if (event.key === "Enter" && document.activeElement.tabIndex === -1) {
      this.props.onClose();
    }
  };

  createChildrenButton = () => {
    let onConfirm;
    let label;
    let tooltip;

    if (Auth.isMasseur() && this.props.event.massage.client === null) {
      return null;
    }

    if (this.props.label === _t.translate("Assign me")) {
      onConfirm = this.onCalendarConfirm;
      label = _t.translate("Assign and add to calendar");
      tooltip = this.props.disabled
        ? _t.translate("Maximal simultaneous massage time per user would be exceeded")
        : _t.translate("Assigns this massage and opens a predefined Google event editor in a new tab");
    } else if (this.props.label === _t.translate("Unassign me")) {
      onConfirm = () => {
        window.open(Util.getEventLink(this.props.event.massage), "_blank");
        this.props.onClose();
      };
      label = _t.translate("Add to Google Calendar");
      tooltip = _t.translate("Opens a predefined Google event editor in a new tab");
    }

    if (label === undefined) {
      return null;
    }
    return (
      <span title={this.props.disabled ? _t.translate("Over the limit") : ""}>
        <Button
          id={this.tooltipTarget}
          className="mr-2"
          tag="a"
          color="primary"
          disabled={this.props.disabled}
          onClick={onConfirm}
          target="_blank"
          rel="noopener noreferrer"
          tabIndex="-1"
        >
          {label}
        </Button>
        <Tooltip
          isOpen={this.state.tooltipActive}
          target={this.tooltipTarget}
          toggle={this.toggleTooltip}
          trigger="hover"
        >
          {tooltip}
        </Tooltip>
      </span>
    );
  };

  createInsides = () => (
    <ModalBody>
      <Row>
        <Col md="12">
          <h3>
            {_t.translate("Details")}
            {(Auth.isAdmin() || (Auth.isMasseur() && Auth.getSub() === this.props.event.massage.masseuse.sub)) && (
              <div className="float-right">
                {this.props.allowEditation && (
                  <TooltipIconButton icon="edit" onClick={this.props.onEdit} tooltip={_t.translate("Edit")} />
                )}
                {this.props.allowDeletion && (
                  <ConfirmationIconButton
                    className="ml-2"
                    icon="trash"
                    onConfirm={this.props.onDelete}
                    tooltip={_t.translate("Delete")}
                  />
                )}
              </div>
            )}
          </h3>
          <hr />
        </Col>
      </Row>

      <Row>
        <Col md="4">
          <dl>
            <dt>{_t.translate("Facility")}</dt>
            <dd className={this.definitionClass}>{this.props.event.massage.facility.name}</dd>
            <dt>{_t.translate("Time")}</dt>
            <dd className={this.definitionClass}>
              {`${moment(this.props.event.massage.date).format("LT")} â€“ ${moment(
                this.props.event.massage.ending
              ).format("LT")}`}
            </dd>
            <dt>{_t.translate("Client")}</dt>
            {Util.isEmpty(this.props.event.massage.client) ? (
              <dd className={this.definitionClass}>
                <p style={{ color: Util.SUCCESS_COLOR }}>
                  <strong>{_t.translate("Free")}</strong>
                </p>
              </dd>
            ) : (
              <dd className={this.definitionClass}>
                <p
                  style={{
                    color:
                      !this.props.allowDeletion ||
                      (this.props.allowEditation && this.props.event.massage.client.sub === Auth.getSub())
                        ? Util.WARNING_COLOR
                        : Util.ERROR_COLOR
                  }}
                >
                  <strong>{this.props.event.massage.client.email}</strong>
                </p>
              </dd>
            )}
          </dl>
        </Col>

        <Col md="4">
          <dl>
            <dt>{_t.translate("Masseur/Masseuse")}</dt>
            <dd className={this.definitionClass}>
              {`${this.props.event.massage.masseuse.name} ${this.props.event.massage.masseuse.surname}`}
            </dd>
            <dt>{_t.translate("Duration")}</dt>
            <dd className={this.definitionClass}>
              {`${moment
                .duration(moment(this.props.event.massage.ending).diff(this.props.event.massage.date))
                .asMinutes()} ${_t.translate("minutes")}`}
            </dd>
          </dl>
        </Col>

        <Col md="4">
          <dl>
            <dt>{_t.translate("Contact")}</dt>
            <dd className={this.definitionClass}>{this.props.event.massage.masseuse.email}</dd>
            <dt>{_t.translate("Date")}</dt>
            <dd className={this.definitionClass}>{moment(this.props.event.massage.date).format("L")}</dd>
          </dl>
        </Col>
      </Row>

      <ModalActions
        primaryLabel={Auth.isMasseur() && this.props.event.massage.client === null ? "none" : this.props.label}
        title={this.props.disabled ? _t.translate("Over the limit") : ""}
        disabled={this.props.disabled}
        onProceed={this.props.onConfirm}
        onClose={this.props.onClose}
      >
        {this.createChildrenButton()}
      </ModalActions>
    </ModalBody>
  );

  render() {
    return this.props.withPortal ? (
      <Modal size="lg" isOpen toggle={this.props.onClose} tabIndex="-1" onKeyPress={this.handleModalKeyPress}>
        {this.createInsides()}
      </Modal>
    ) : (
      this.createInsides()
    );
  }
}

MassageEventModal.propTypes = {
  /** event to be displayed */
  event: PropTypes.shape({
    bgColor: PropTypes.string,
    massage: PropTypes.shape({
      id: PropTypes.number,
      masseuse: PropTypes.shape({
        email: PropTypes.string,
        masseur: PropTypes.bool,
        name: PropTypes.string,
        sub: PropTypes.string,
        subscribed: PropTypes.bool,
        surname: PropTypes.string
      }),
      date: PropTypes.oneOfType([PropTypes.number, PropTypes.instanceOf(Date)]),
      ending: PropTypes.oneOfType([PropTypes.number, PropTypes.instanceOf(Date)]),
      client: PropTypes.shape({
        email: PropTypes.string,
        masseur: PropTypes.bool,
        name: PropTypes.string,
        surname: PropTypes.string,
        sub: PropTypes.string,
        subscribed: PropTypes.bool
      }),
      facility: PropTypes.shape({
        id: PropTypes.number,
        name: PropTypes.string
      })
    })
  }).isRequired,
  /** primary button label */
  label: PropTypes.string.isRequired,
  /** function called on modal dismissal */
  onClose: PropTypes.func.isRequired,
  /** function called on primary label action */
  onConfirm: PropTypes.func.isRequired,
  /** whether the delete button should be shown */
  allowDeletion: PropTypes.bool,
  /** whether the edit button should be shown */
  allowEditation: PropTypes.bool,
  /** whether the primary button should be disabled */
  disabled: PropTypes.bool,
  /** function called on edit button action */
  onEdit: PropTypes.func,
  /** function called on delete button action */
  onDelete: PropTypes.func,
  /** whether ModalContainer should be used; useful for testing to avoid portals */
  withPortal: PropTypes.bool
};

MassageEventModal.defaultProps = {
  allowDeletion: true,
  allowEditation: true,
  disabled: false,
  onEdit() {},
  onDelete() {},
  withPortal: true
};

export default MassageEventModal;
