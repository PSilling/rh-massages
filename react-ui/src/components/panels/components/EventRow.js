// react imports
import React, { Component } from "react";
import PropTypes from "prop-types";

// module imports
import moment from "moment";

// component imports
import MassageEventModal from "../../modals/MassageEventModal";

// util imports
import _t from "../../../util/Translations";
import Auth from "../../../util/Auth";
import Util from "../../../util/Util";

/**
 * Custom component event for MyMassages agenda calendar view.
 */
class EventRow extends Component {
  state = { active: false };

  handleToggle = () => {
    this.setState(prevState => ({ active: !prevState.active }));
  };

  handleKeyPress = event => {
    if (event.key === "Enter") {
      this.handleToggle();
    }
  };

  getUserInfo = () => {
    if (Auth.isMasseur()) {
      if (this.props.event.massage.client === null) {
        return <span style={{ color: Util.SUCCESS_COLOR }}>{_t.translate("Free")}</span>;
      }
      return (
        <span style={{ color: Util.ERROR_COLOR }}>
          {`${this.props.event.massage.client.name} ${this.props.event.massage.client.surname}`}
        </span>
      );
    }

    return <span>{`${this.props.event.massage.masseuse.name} ${this.props.event.massage.masseuse.surname}`}</span>;
  };

  render() {
    const dateString = moment(this.props.event.massage.date).format("L");
    return (
      <tr>
        <td>{this.props.event.massage.facility.name}</td>
        <td>{dateString.substring(0, dateString.length - 5)}</td>
        <td>
          {`${moment(this.props.event.massage.date).format("LT")} â€“ ${moment(this.props.event.massage.ending).format(
            "LT"
          )}`}
        </td>
        <td colSpan="2">
          <strong
            onClick={this.handleToggle}
            role="button"
            style={{ cursor: "pointer" }}
            tabIndex="-1"
            onKeyPress={this.handleKeyPress}
          >
            {`${_t.translate("Massage")}: `}
            {this.getUserInfo()}
          </strong>
          <br />

          {this.state.active && (
            <MassageEventModal
              event={this.props.event}
              label={Auth.isMasseur() ? _t.translate("Force cancel") : _t.translate("Unassign me")}
              allowEditation={false}
              allowDeletion={false}
              onClose={this.handleToggle}
              onConfirm={() => {
                this.handleToggle();
                this.props.onCancel(this.props.event.massage);
              }}
            />
          )}
        </td>
      </tr>
    );
  }
}

EventRow.propTypes = {
  /** event to be rendered */
  event: PropTypes.shape({
    bgColor: PropTypes.string,
    massage: PropTypes.shape({
      id: PropTypes.number,
      masseuse: PropTypes.shape({
        email: PropTypes.string,
        masseur: PropTypes.bool,
        name: PropTypes.string,
        sub: PropTypes.string,
        subscribed: PropTypes.bool,
        surname: PropTypes.string
      }),
      date: PropTypes.oneOfType([PropTypes.number, PropTypes.instanceOf(Date)]),
      ending: PropTypes.oneOfType([PropTypes.number, PropTypes.instanceOf(Date)]),
      client: PropTypes.shape({
        email: PropTypes.string,
        name: PropTypes.string,
        surname: PropTypes.string,
        sub: PropTypes.string,
        subscribed: PropTypes.bool
      }),
      facility: PropTypes.shape({
        id: PropTypes.number,
        name: PropTypes.string
      })
    })
  }).isRequired,
  /** function called on massage cancellation */
  onCancel: PropTypes.func.isRequired
};

export default EventRow;
