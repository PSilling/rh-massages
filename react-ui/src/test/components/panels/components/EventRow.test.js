// react imports
import React from "react";
import { shallow } from "enzyme";

// test imports
import MassageEventModal from "../../../../components/modals/MassageEventModal";
import EventRow from "../../../../components/panels/components/EventRow";

test("renders content with correct props", () => {
  const testFunction = jest.fn();
  const testEvent = {
    massage: {
      id: 1,
      date: new Date(0),
      ending: new Date(1000),
      masseuse: "test",
      client: null,
      facility: { id: 1, name: "test" }
    }
  };
  const wrapper = shallow(<EventRow event={testEvent} onCancel={testFunction} />);
  const strongButton = wrapper.find("strong");
  let modals = wrapper.find(MassageEventModal);

  expect(modals.length).toBe(0);
  strongButton.props().onClick();
  modals = wrapper.find(MassageEventModal);
  expect(modals.length).toBe(1);
  expect(modals.props().event).toBe(testEvent);
  expect(testFunction).not.toHaveBeenCalled();

  modals.props().onConfirm();
  modals = wrapper.find(MassageEventModal);
  expect(modals.length).toBe(0);
  expect(testFunction).toHaveBeenCalledTimes(1);

  expect(wrapper).toMatchSnapshot();
});
