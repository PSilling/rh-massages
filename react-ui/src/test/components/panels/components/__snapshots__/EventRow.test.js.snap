// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders content with correct props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EventRow
    event={
      Object {
        "massage": Object {
          "client": null,
          "date": 1970-01-01T00:00:00.000Z,
          "ending": 1970-01-01T00:00:01.000Z,
          "facility": Object {
            "id": 1,
            "name": "test",
          },
          "id": 1,
          "masseuse": Object {
            "email": "test@masseuse.org",
            "masseur": true,
            "name": "Masseuse",
            "sub": "m-sub",
            "subscribed": false,
            "surname": "Test",
          },
        },
      }
    }
    onCancel={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "client": null,
              "date": 1970-01-01T00:00:00.000Z,
              "ending": 1970-01-01T00:00:01.000Z,
              "facility": Object {
                "id": 1,
                "name": "test",
              },
              "id": 1,
              "masseuse": Object {
                "email": "test@masseuse.org",
                "masseur": true,
                "name": "Masseuse",
                "sub": "m-sub",
                "subscribed": false,
                "surname": "Test",
              },
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <td>
          01/01
        </td>,
        <td>
          1:00 AM – 1:00 AM
        </td>,
        <td
          colSpan="2"
        >
          <strong
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            style={
              Object {
                "cursor": "pointer",
              }
            }
            tabIndex="-1"
          >
            Masáž v test
          </strong>
          <br />
          <span>
            Masseuse Test
          </span>
        </td>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "01/01",
        },
        "ref": null,
        "rendered": "01/01",
        "type": "td",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "1:00 AM – 1:00 AM",
        },
        "ref": null,
        "rendered": "1:00 AM – 1:00 AM",
        "type": "td",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <strong
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              style={
                Object {
                  "cursor": "pointer",
                }
              }
              tabIndex="-1"
            >
              Masáž v test
            </strong>,
            <br />,
            <span>
              Masseuse Test
            </span>,
            false,
          ],
          "colSpan": "2",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Masáž v test",
              "onClick": [Function],
              "onKeyPress": [Function],
              "role": "button",
              "style": Object {
                "cursor": "pointer",
              },
              "tabIndex": "-1",
            },
            "ref": null,
            "rendered": "Masáž v test",
            "type": "strong",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Masseuse Test",
            },
            "ref": null,
            "rendered": "Masseuse Test",
            "type": "span",
          },
          false,
        ],
        "type": "td",
      },
    ],
    "type": "tr",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <td>
            01/01
          </td>,
          <td>
            1:00 AM – 1:00 AM
          </td>,
          <td
            colSpan="2"
          >
            <strong
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              style={
                Object {
                  "cursor": "pointer",
                }
              }
              tabIndex="-1"
            >
              Masáž v test
            </strong>
            <br />
            <span>
              Masseuse Test
            </span>
          </td>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "01/01",
          },
          "ref": null,
          "rendered": "01/01",
          "type": "td",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "1:00 AM – 1:00 AM",
          },
          "ref": null,
          "rendered": "1:00 AM – 1:00 AM",
          "type": "td",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <strong
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
                tabIndex="-1"
              >
                Masáž v test
              </strong>,
              <br />,
              <span>
                Masseuse Test
              </span>,
              false,
            ],
            "colSpan": "2",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Masáž v test",
                "onClick": [Function],
                "onKeyPress": [Function],
                "role": "button",
                "style": Object {
                  "cursor": "pointer",
                },
                "tabIndex": "-1",
              },
              "ref": null,
              "rendered": "Masáž v test",
              "type": "strong",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Masseuse Test",
              },
              "ref": null,
              "rendered": "Masseuse Test",
              "type": "span",
            },
            false,
          ],
          "type": "td",
        },
      ],
      "type": "tr",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
